caiman/motion_correction.py

```diff
@@ -2464,6 +2497,117 @@ def tile_and_correct_3d(img:np.ndarray, template:np.ndarray, strides:tuple, over
+        shifts_x = cv2.resize(shift_img_x.astype(np.float32), dims[::-1])
+    
+    # apply to image
+    if border_nan is False:
+        mode = cv2.BORDER_CONSTANT
+        value = 0.0
+    elif border_nan is True:
+        mode = cv2.BORDER_CONSTANT
+        value = np.nan
+    elif border_nan == 'min':
+        mode = cv2.BORDER_CONSTANT
+        value = np.min(img)
+    elif border_nan == 'copy':
+        mode = cv2.BORDER_REPLICATE
+        value = 0.0
+    else:
+        raise ValueError(f'Unknown value of border_nan ({border_nan})')
+    
+    return cv2.remap(img, shifts_x + x_grid, shifts_y + y_grid, cv2.INTER_CUBIC,
+                     borderMode=mode, borderValue=value)
+
+def apply_pw_shifts_remap_3d(img: np.ndarray, shifts_y: np.ndarray, shifts_x: np.ndarray, shifts_z: np.ndarray,
+                             patch_centers: tuple[list[float], ...], border_nan: Union[bool, Literal['copy', 'min']],
+                             interp_shifts_precisely=False) -> np.ndarray:
+    """
+    Use skimage warp to apply 3D piecewise shifts
+    Inputs:
+        img: the 3D image to apply shifts to
+        shifts_y: array of y shifts for each patch (C order) (this is the actual Y i.e. the first dimension of the image)
+        shifts_x: array of x shifts for each patch (C order)
+        shifts_z: array of z shifts for each patch (C order)
+        patch_centers: tuple of patch locations in each dimension.
+        border_nan: how to deal with borders when remapping
+        interp_shifts_precisely: if true, uses interpn to upsample shifts based on patch centers instead of resize
+    Outputs:
+        img_remapped: the remapped image
+    """
+    # reshape shifts based on patch grid
+    patch_grid = tuple(len(centers) for centers in patch_centers)
+    shift_img_y = np.reshape(shifts_y, patch_grid)
+    shift_img_x = np.reshape(shifts_x, patch_grid)
+    shift_img_z = np.reshape(shifts_z, patch_grid)
+
+    # get full image shape/coordinates
+    dims = img.shape
+    x_coords, y_coords, z_coords = [np.arange(0., dims[dim]).astype(np.float32) for dim in (1, 0, 2)]
+    x_grid, y_grid, z_grid = np.meshgrid(x_coords, y_coords, z_coords)
+
+    # up-sample shifts
+    if interp_shifts_precisely:
+        coords_new = (y_coords, x_coords, z_coords)
+        shifts_y = interpolate_shifts(shift_img_y, patch_centers, coords_new).astype(np.float32)
+        shifts_x = interpolate_shifts(shift_img_x, patch_centers, coords_new).astype(np.float32)
+        shifts_z = interpolate_shifts(shift_img_z, patch_centers, coords_new).astype(np.float32)
+    else:
+        shifts_y = resize_sk(shift_img_y.astype(np.float32), dims)
+        shifts_x = resize_sk(shift_img_x.astype(np.float32), dims)
+        shifts_z = resize_sk(shift_img_z.astype(np.float32), dims)
+    
+    shift_map = np.stack((shifts_y + y_grid, shifts_x + x_grid, shifts_z + z_grid), axis=0)
+    
+    # apply to image
+    if border_nan is False:
```

Comment:
if not border_nan:

Reply:
technically this would work, but given the fact that it's a weird type I feel like it's better to be more explicit.

Reply:
Fair enough.